### PR TITLE
Allow 0 replicas

### PR DESCRIPTION
**Pull request by michaelgugino**
This is needed for BYO team to have 0 worker replicas.

### CODE REVIEW

**Comment by staebler** [comment]
Wouldn't it be better to get rid of the validation that requires the worker machine pool?

File: pkg/types/validation/machinepools.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ func ValidateMachinePool(p *types.MachinePool, fldPath *field.Path, platform str
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("name"), p.Name, validMachinePoolNameValues))
 	}
 	if p.Replicas != nil {
-		if *p.Replicas <= 0 {
+		if *p.Replicas < 0 {
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("replicas"), p.Replicas, "number of replicas must be positive"))
```
Comment (staebler) [inline pkg/types/validation/machinepools.go]: Change the text in the error generated to reflect that the number of replicas must not be negative.

File: pkg/types/validation/machinepools.go [hunk 2]
```diff
@@ -38,7 +38,7 @@ func ValidateMachinePool(p *types.MachinePool, fldPath *field.Path, platform str
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("name"), p.Name, validMachinePoolNameValues))
 	}
 	if p.Replicas != nil {
-		if *p.Replicas <= 0 {
+		if *p.Replicas < 0 {
```
Comment (michaelgugino) [inline pkg/types/validation/machinepools.go]: I'm not sure why this is necessary of all a sudden.  It worked fine in v0.6.0.  Does not work in v0.7.0-master or master.

Comment (staebler) [reply pkg/types/validation/machinepools.go]: Validation was added in v0.7.0.

**Comment by staebler** [comment]
This will make it valid to set a replica size of 0 for the master machine pool as well. I don't think we want that.

**Comment by michaelgugino** [comment]
> This will make it valid to set a replica size of 0 for the master machine pool as well. I don't think we want that.

@staebler  I can make it check just masters if you prefer.

**Comment by staebler** [comment]
> > This will make it valid to set a replica size of 0 for the master machine pool as well. I don't think we want that.
> 
> @staebler I can make it check just masters if you prefer.

@michaelgugino My preference is to remove the requirement for the "worker" machine pool. Is there a reason to keep that machine pool with a replica size of 0 instead of omitting that machine pool?

**Comment by michaelgugino** [comment]
> > > This will make it valid to set a replica size of 0 for the master machine pool as well. I don't think we want that.
> > 
> > 
> > @staebler I can make it check just masters if you prefer.
> 
> @michaelgugino My preference is to remove the requirement for the "worker" machine pool. Is there a reason to keep that machine pool with a replica size of 0 instead of omitting that machine pool?

@staebler 

Aside from the fact that there are other validations in place expect that to be present?  No, I don't think we need it to be present at all, but that would be a bigger patch set.  Keeping it also would result in the machineset being created, which might be desired for scaling later in other usecases, of which I don't have a compelling example ATM.

While we could exclude workers, I think it makes more sense to specifically include masters.  I think in all likelihood there will be an expectation to have arbitrary machinesets in the future, but it makes no difference to me.

Let me know what you prefer, I think the simplest fix is the best here, I just need it to not fail when I want to create ignition configs with 0 workers.

**Comment by staebler** [comment]
> > > > This will make it valid to set a replica size of 0 for the master machine pool as well. I don't think we want that.
> > > 
> > > 
> > > @staebler I can make it check just masters if you prefer.
> > 
> > 
> > @michaelgugino My preference is to remove the requirement for the "worker" machine pool. Is there a reason to keep that machine pool with a replica size of 0 instead of omitting that machine pool?
> 
> @staebler
> 
> Aside from the fact that there are other validations in place expect that to be present? No, I don't think we need it to be present at all, but that would be a bigger patch set. Keeping it also would result in the machineset being created, which might be desired for scaling later in other usecases, of which I don't have a compelling example ATM.
> 
> While we could exclude workers, I think it makes more sense to specifically include masters. I think in all likelihood there will be an expectation to have arbitrary machinesets in the future, but it makes no difference to me.
> 
> Let me know what you prefer, I think the simplest fix is the best here, I just need it to not fail when I want to create ignition configs with 0 workers.

We can do either, especially for the short term. I think changing the validation to remove the "worker" requirement is easy, too. Delete these 3 lines. https://github.com/openshift/installer/blob/ac006ae671a645553d58c8a29c676968dfa3d85f/pkg/types/validation/installconfig.go#L103-L105

**Comment by staebler** [comment]
@michaelgugino Whichever you prefer to do in the short term, I will lgtm to get BYO unblocked on this.

**Comment by michaelgugino** [comment]
@staebler ptal

**Comment by sdodson** [comment]
/cc @wking @abhinavdahiya 
/approve
PTAL

lgtm

**Comment by crawford** [comment]
@michaelgugino I ran into this recently as well. I think I'd prefer @staebler's suggestion of removing the need for the worker machine pool. Do you have time to take another pass at this? If not, I'll eventually pick it up.

**Comment by michaelgugino** [comment]
> @michaelgugino I ran into this recently as well. I think I'd prefer @staebler's suggestion of removing the need for the worker machine pool. Do you have time to take another pass at this? If not, I'll eventually pick it up.

@crawford  I think that would still require this patch set.  While the workers could indeed not be defined, if they are defined and set to zero, we'd still hit the same error.  IMO, allowing for zero makes more sense as it's easier to template.

**Comment by wking** [comment]
On the operator-support front, I've [filed Bugzilla issues for the three remaining pods which do not tolerate masters][1]:

* Ingress: https://bugzilla.redhat.com/show_bug.cgi?id=1671136
* Prometheus: https://bugzilla.redhat.com/show_bug.cgi?id=1671137
* OLM: https://bugzilla.redhat.com/show_bug.cgi?id=1671140

[1]: https://github.com/openshift/installer/pull/1146#issuecomment-459102371

**Comment by crawford** [comment]
Superseded by https://github.com/openshift/installer/pull/1231.

**Pull request closed without merging**